### PR TITLE
feat: adds user role support

### PR DESCRIPTION
**Pull request by jscyo**
## Summary of change
Adds UserRoles support 

## Related issues
- https://github.com/supertokens/supertokens-core/pull/414

## Test Plan
Run core tests

## Documentation changes
- none

## Checklist for important updates
- [x] Changelog has been updated
- [x] `pluginInterfaceSupported.json` file has been updated (if needed)
- [x] Changes to the version if needed
   - In `build.gradle`
- [x] Had installed and ran the pre-commit hook
- [ ] ~~If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.~~
- [x] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- none

### CODE REVIEW

File: src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            pst.setString(1, userId);
+            pst.setString(2, role);
+        });
+    }
+
+    public static boolean createNewRoleOrDoNothingIfExists_Transaction(Start start, Connection con, String role)
+            throws SQLException, StorageQueryException {
+        // We use IGNORE here since we want to ignore any conflict(duplicate key exception) that might be thrown.
+        // We do not use ON DUPLICATE UPDATE since it will return false positives when a role already exists
+        String QUERY = "INSERT IGNORE INTO " + Config.getConfig(start).getRolesTable() + " VALUES(?)";
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java:83]: I don't think using `INSERT IGNORE` is good. Cause it allows you to insert null values into a non null column (based on https://stackoverflow.com/a/548570/3867175 answer)

Instead, you should use `... ON DUPLICATE KEY UPDATE` method. According to this answer (https://stackoverflow.com/a/7853634/3867175) it still shows number of rows affected as 0.

What do you mean by "return false positives"?

File: src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    public static boolean deleteRole(Start start, String role) throws SQLException, StorageQueryException {
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getRolesTable() + " WHERE role = ? ;";
+        return update(start, QUERY, pst -> {
+            pst.setString(1, role);
+        }) == 1;
+    }
+
+    public static boolean doesRoleExist_transaction(Start start, Connection con, String role)
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java:212]:
```suggestion
    public static boolean doesRoleExist_Transaction(Start start, Connection con, String role)
```

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java:212]: All other places, the `T` is capital. So keep this consistent

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java [hunk 3]
```diff
@@ -78,10 +77,17 @@ public static void addRoleToUser(Start start, String userId, String role)
...
-        // We do not use ON DUPLICATE UPDATE since it will return false positives when a role already exists
-        String QUERY = "INSERT IGNORE INTO " + Config.getConfig(start).getRolesTable() + " VALUES(?)";
-        int rowsUpdated = update(con, QUERY, pst -> pst.setString(1, role));
+        // We want to insert a role which does not exist into the roles table, Since INSERT queries do not support WHERE
+        // condition clause, we use a SELECT query with the input role and a WHERE NOT EXISTS clause to check that the
+        // input role does not exist in the table
+
+        String QUERY = "INSERT INTO " + Config.getConfig(start).getRolesTable() + "(role) "
+                + "SELECT ? AS role WHERE NOT EXISTS (" + " SELECT role FROM " + Config.getConfig(start).getRolesTable()
+                + " WHERE role = ? )";
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java:86]: Is this atmoic?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java (hunk 3 above)
Comment (jscyo) [reply src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java:86]: So according to mysql docs [here](https://dev.mysql.com/doc/internals/en/transactions-historical-note.html), Mysql starts a transaction for every sql query, so the subqueries are also wrapped in a transaction. According to MySQL [docs](https://dev.mysql.com/doc/refman/8.0/en/innodb-transaction-isolation-levels.html) description of serializable, they convert regular SELECT to SELECT...FOR SHARE which would prevent the row from being deleted while this transaction is going on

**Pull request merged**